src/app/beer_garden/config.py

```diff
@@ -0,0 +1,1042 @@
+    config = spec.load_filtered_config(cli_vars, "ENVIRONMENT", exclude_bootstrap=True)
+
+    dump_data(config, filename=bootstrap.configuration.file, file_type="yaml")
+
+
+def migrate(args: Sequence[str]):
+    """Updates a configuration file in-place.
+
+    Args:
+        args: Command line arguments. Must contain an argument that specifies the
+        config file to update ('-c')
+    Returns:
+        None
+
+    Raises:
+        YapconfLoadError: Missing 'config' configuration option (file location)
+    """
+    spec, cli_vars = _parse_args(args)
+
+    config = spec.load_config(cli_vars, "ENVIRONMENT")
+
+    if not config.configuration.file:
+        raise SystemExit(
+            "Please specify a config file to update" " in the CLI arguments (-c)"
+        )
+
+    current_root, current_extension = os.path.splitext(config.configuration.file)
+
+    current_type = current_extension[1:]
+    if current_type == "yml":
+        current_type = "yaml"
+
+    # Determine if a type conversion is needed
+    type_conversion = False
+    new_type = "yaml"
+    if current_type != new_type:
+        new_file = current_root + "." + new_type
+        type_conversion = True
+    else:
+        new_file = config.configuration.file
+
+    # logger.debug("About to migrate config at %s" % config.configuration.file)
+    spec.migrate_config_file(
+        config.configuration.file,
+        current_file_type=current_type,
+        output_file_name=new_file,
+        output_file_type=new_type,
+        update_defaults=True,
+        include_bootstrap=False,
+    )
+
+    if type_conversion:
+        # logger.debug("Removing old config file at %s" % config.configuration.file)
+        os.remove(config.configuration.file)
+
+
+def generate_logging(args: Sequence[str]):
+    """Generate and save logging configuration file.
+
+    Args:
+        args: Command line arguments
+            --log-config-file: Configuration will be written to this file (will print to
+                stdout if missing)
+            --log-file: Logs will be written to this file (used in a RotatingFileHandler)
+            --log-level: Handlers will be configured with this logging level
+
+    Returns:
+        str: The logging configuration dictionary
+    """
+    spec, cli_vars = _parse_args(args)
+    filtered_args = spec.load_filtered_config(
+        cli_vars, include=["log.config_file", "log.file", "log.level"]
+    )
+
+    log = filtered_args.get("log", {})
+    logging_config = default_app_config(log.get("level"), log.get("file"))
+    log_config_file = log.get("config_file")
+
+    if log_config_file is not None:
+        with open(log_config_file, "w") as f:
+            dumped = json.dumps(logging_config, indent=4, sort_keys=True)
+            f.write(six.u(dumped))
+    else:
+        print(json.dumps(logging_config, indent=4, sort_keys=True))
+
+    return logging_config
+
+
+def get(key: str):
```

Comment:
One thing I was thinking about this function call. We are still opening ourselves up to allowing people to change the config. Do you think we should be returning an immutable copy of the config?

Reply:
For example:

```
config.load()
log_config = config.get("log")
log_config["level"] = "this probably should not be allowed"
```

Reply:
Yeah, I'm good with that.

Reply:
Do you think this should be done at load time? We can do something like this:

```python
_CONFIG = Box(spec.load_config(*config_sources).to_dict(), frozen_box=True)
```

Reply:
Yep! I think that's great!